### PR TITLE
EZP-30460: Remove deprecated \eZ\Publish\Core\FieldType\FieldType::getName method

### PR DESCRIPTION
**Pull request by mikadamczyk**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30460](https://jira.ez.no/browse/EZP-30460)
| **Improvement**| yes
| **New feature**    | no
| **Target version** | `master`
| **BC breaks**      |no
| **Tests pass**     | yes/no
| **Doc needed**     | yes/no

**TODO**:
- [ ] Implement feature / fix a bug.
- [ ] Implement tests.
- [ ] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [ ] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/FieldType/Page/NameableField.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\FieldType\Page;
+
+use eZ\Publish\API\Repository\Values\ContentType\FieldDefinition;
+use eZ\Publish\SPI\FieldType\Nameable;
+use eZ\Publish\SPI\FieldType\Value;
+
+class NameableField implements Nameable
```
Comment (adamwojs) [inline eZ/Publish/Core/FieldType/Page/NameableField.php]: `ezpage` field type has been dropped so this is no longer needed.

File: eZ/Publish/SPI/FieldType/FieldType.php [hunk 2]
```diff
@@ -43,21 +43,6 @@ interface FieldType
...
-     * is designated to be used in the content name/urlAlias pattern.
-     *
-     * The used $value can be assumed to be already accepted by {@link * acceptValue()}.
-     *
-     * @deprecated Since 6.3/5.4.7, use \eZ\Publish\SPI\FieldType\Nameable
-     * @param \eZ\Publish\SPI\FieldType\Value $value
-     *
-     * @return string
-     */
-    public function getName(Value $value);
```
Comment (adamwojs) [inline eZ/Publish/SPI/FieldType/FieldType.php]: This change should be mentioned in  `/doc/bc/changes-8.0.md`

File: eZ/Publish/Core/FieldType/Page/NameableField.php (hunk 1 above)
Comment (mikadamczyk) [reply eZ/Publish/Core/FieldType/Page/NameableField.php]: If PR will be merged this should be removed

**Comment by webhdx** [comment]
I don't like the direction here - it's too complicated. We don't need `Nameable` interface anymore. Just add this method as a part of FieldType interface since it's required. Then you can avoid all this magic happening in compiler pass.

**Comment by andrerom** [comment]
> We don't need Nameable interface anymore.

How is name on [selection/relation fields](https://github.com/ezsystems/ezpublish-kernel/pull/1605) handled now to avoid it?  Or do you imply merging the method into main interface?

**Comment by webhdx** [comment]
@andrerom I was thinking about merging it to the fieldtype interface. Even though it's not mandatory we still provide fallback for all non nameable fieldtypes which suggests it should be required and the "fallback" solution could be a part of generic fieldtype.

**Comment by andrerom** [comment]
👍 from me, just ideally make sure partners can somehow have field types working across 7.x and 8.x, hence why I suggested deprecating the Nameable interface when moving method (with then exact same signature) to main interface. Like Symfony usuallyy [does](https://symfony.com/doc/current/contributing/code/bc.html).

**Comment by alongosz** [comment]
I fully agree with @webhdx. This would also unblock #2611 providing less confusion there.
We probably should deprecate Nameable, but in 2.5.x.

**Comment by andrerom** [comment]
> We probably should deprecate Nameable, but in 2.5.x.

And how do you expect to accomplish forward compatibility then? 😏 
ref: `have field types working across 7.x and 8.x`

**Comment by alongosz** [comment]
> > We probably should deprecate Nameable, but in 2.5.x.
> 
> And how do you expect to accomplish forward compatibility then?  _(assuming you imply removing it in 8.x, and not providing analternativee in 7.5 as it would be BC break)_ 

One idea: what implements now Field Type interface should also implement Nameable. In 3.0 Nameable can be removed by simple usages search as Field Type interface will require method from Nameable.

> ref: `have field types working across 7.x and 8.x`

Let's stop supporting everything in every version, this is huge maintenance cost and blocker for productivity and progress. I need to see real difficulty in upgrade process to be softer on changes. I don't think that usage search is a difficulty, it happens with every major version of every software following semver. Moreover I don't intend to support people who use notepad to work on a project.

This is of course still open for a debate depending on final upgrade steps needed. Field Types are one of the key components of our system, so naturally I'd like to make migration as painless as possible. Common sense applied ;)

**Comment by gggeek** [comment]
@alongosz "Let's stop supporting everything in every version"

Sorry, a big -1 from me on this one.

As maintainer of a few extensions, I absolutely do not want to have to be forced to update my own code in 
lockstep with the major releases of the core platform.
I do work, today, with customers that use eg. eZPublish 5.4, and I am happy that I can use for them the same extensions that other developers in the community can use in projects based on eZPlatform 2.
If you start allowing API changes with every major version, that means that people like me will have to maintain separate branches of their open-source extensions, one for each kernel version. Given the fact that we are doing it either for free or in our spare time, and that for each extension there are 1, maybe 2 core devs doing most of the work, this will most likely than not mean fewer community extensions available.

In other words: please don't adopt the bad practices from other, less professional CMS. If you do so, you might find that developers have few reasons to stick with your one... ;-)

**Comment by lserwatka** [comment]
@gggeek we haven't done any changes in the FieldType API since several years. Given the fact that eZ Platform 3.0 will use Symfony Flex, which by default forces you as maintainer to adopt your code base to new format, you will have to maintain different version of your code for 2.x and 3.x. We want to use this opportunity to clean this API which is a mess at the moment. It has nothing to do with adopting bad practices, I found this a bit unfair as for many years we had solid API without major breaks. Since Symfony Flex architecture enforces us developers to adopt our codebase we want to use use this opportunity to clean up as much as we can. We can't also predict all changes and deprecate it upfront. This is unrealistic, we are doing this agile way.

**Comment by gggeek** [comment]
I don't want to hijack this PR with an OT discussion, as I seem to have an innate tendency to do, but I still politely disagree.
Eg, afaik, there is no need for a bundle to be rewritten to be compatible with a flex app. It's only the dir structure of the app itself that needs change.
As for adopting agile: I am all in favour of agile for managing dev teams and end-projects, but nothing I have seen so far has convinced me that it is a sensible choice for managing public apis.
I'll be happy to carry further the discussion in person or anywhere more appropriate online :-)

**Comment by lserwatka** [comment]
I want to stick to the facts. From eZ Publish 5.4 up to 2.x we have not done any changes to the Public API which could break your software. As you pointed out, for several years you could reuse FieldTypes across different versions. We have a technical debt that has been accumulated for past several years and this is good opportunity to make cleanups, because of huge changes for Symfony 4. With eZ Platform 3.0 we will do some breaks for sure and likely changes will require developer to change one line, maybe two, taking this example, you will likely change from `implements` to `extends`. This will open door for us for more flexibility in future. Given you have to re-do your bundle structure for Flex, changing two extra lines is not a big deal, semantically you will have to bump your master version anyway and update composer.json requirements for eZ Platform 3.0. We don't apply agile principle for Public APIs as you could miss-understand.

**Comment by andrerom** [comment]
I don't think @gggeek is telling we shouldn't do changes. It's about the reasoning he objects to I believe. And Semantic version wise, ideally we should have deprecated and provided future proof way of writing Fieldtypes for 2.5 before it was out. Allowing code _(disregarding directory changes with flex)_ to work across 2.5 and 3.0 once out.

As we don't release 2.5 and 3.0 at the same time, we can do as we done before, backport deprecation and the forward compatibility.  Can we accomplish that?

**Comment by mikadamczyk** [comment]
Closed in favor of [https://github.com/ezsystems/ezpublish-kernel/pull/2661](https://github.com/ezsystems/ezpublish-kernel/pull/2661)
